### PR TITLE
Migrate to Gulp & Integrate Mock (v2.0.0)

### PR DESCRIPTION
**Pull request by markhayden**
### WHEN REVIEWING MAKE SURE YOU MARK UP FILES IN /SRC NOT /OUT

Original PR: https://github.com/lytics/jstag/pull/21
First step in actually closing: https://github.com/lytics/lio/issues/6881

Once the `delete-junk` branch gets thumbed / merged we can probably alter this pr to be against gh-pages but for now this will make it easier to review the code related changes. Also plan on rebasing this thing this afternoon.

`out` are the compiled files so all comments should be against files in `src`

## Cleanup
Migrated everything to use gulp.

## Tests
Added gulp, jasmine, karma, phantom for unit tests. Wrote tests for everything I could think of. I am sure ther are some missing but here is a shortlist of the things that are covered:

* extend function
* setcookie
* getcookie
* async queue
* send
* identify
* pageview
* mock
* block
* etc

## New Functionality
In support of 6881 I introduced the concept of a `mock` and a `hold` on sending events. We needed the ability to basically pause all sends and then build an object of that send queued data so that we could pass it to our `/me` endpoint and get the most current state of the entity. This will make strides in facilitating the `first visit` use case as well as many other use cases negatively effected by latency.

In a nutshell it routes all send events to a queue. Then when send the mock event it pulls all that data out of the queue, extends the current payload and does all the processing up the point where it would actually be sent. It then returns the payload via callback so we can handle it in `/tag/lio.js`.

When block is lifted all existing events get processed normally.

### TODO:
* Add Readme
* Rebase all Commits

### CODE REVIEW

File: src/io.js [hunk 1]
```diff
@@ -316,10 +322,35 @@
...
+
+      for (var i = window.jstag._q.length - 1; i >= 0; i--) {
+        event = window.jstag._q[i];
+
+        // since we throw event level streams out for mock just get last in array
+        if(event.length === 3){
+          data = event[2][event[2].length-1];
+        }
+
+        merged = extend(merged, data);
```
Comment (DingoEatingFuzz) [inline src/io.js]: In the case where `event` doesn't have a length of three, `data` will either be undefined (if this is the first run of the loop) or the value of the last good event. In either case I think it results in a no-op, but it would be cleaner to move the `extend` operation to the conditional.

```js
if (event.length === 3) {
  merged = extend(merged, event[2][event[2].length-1]);
}
```

This also alleviates the need for the `data` variables altogether.

Comment (DingoEatingFuzz) [inline src/io.js]: It also looks like this `mergeTempQ` function is never used.

Comment (markhayden) [inline src/io.js]: yah this can all go. it was a first attempt before realizing we needed to block the sending of events at the jstag.send level.

**Comment by DingoEatingFuzz** [comment]
👍 × 💯

**Comment by araddon** [comment]
:+1:
